File: C/Mergesort.c
/*
    Input : size of array and the array 
    Output: Sorted Array
    Algorithm: Mergesort
    Time Complexity: O(n*log n)
    Space Complexity: O(n)
    
    Sample Input and Output:
    Enter the size of the array :5
    Enter the array elements :
     5 4 1 2 3
    Sorted Array :  1 2 3 4 5
    
*/



#include<stdio.h>
#include<stdlib.h>

void merge(int arr[], int l, int m, int r) 
{ 
    int i, j, k; 
    int n1 = m - l + 1; 
    int n2 =  r - m;    
    int L[n1], R[n2];   // Taking temp arrays  left and right 
    for (i = 0; i < n1; i++) 
        L[i] = arr[l + i]; 
    for (j = 0; j < n2; j++) 
        R[j] = arr[m + 1+ j];   
    i = 0; 
    j = 0; 
    k = l; 
    while (i < n1 && j < n2)  
    { 
        if (L[i] <= R[j]) 
        { 
            arr[k] = L[i]; 
            i++; 
        } 
        else
        { 
            arr[k] = R[j]; 
            j++; 
        } 
        k++; 
    } 
      while (i < n1) 
    { 
        arr[k] = L[i]; 
        i++; 
        k++; 
    } 
      while (j < n2) 
    { 
        arr[k] = R[j]; 
        j++; 
        k++; 
    } 
} 
  void mergeSort(int arr[], int l, int r) // Recursive Merge Sort
{ 
    if (l < r) 
    { 
        
        int m = l+(r-l)/2;  
       
        mergeSort(arr, l, m); // First Half
        mergeSort(arr, m+1, r);   // Last Half
        merge(arr, l, m, r); // Marge
    } 
}
int main(){
    int *arr,n,i;

    printf("Enter the size of the array :");
    scanf("%d",&n);
    printf("Enter the array elements :\n ");
    for(i=0;i<n;i++){
        scanf("%d",&arr[i]);
    }
    mergeSort(arr,0,n-1);
    printf("Sorted Array :\t");
    for(i=0;i<n;i++){
        printf("%d ",arr[i]);
    }


}

File: Python/mergeSort.py
# Python program for implementation of MergeSort

# Merge two subarrays of arr[].
# 1st subarray is arr[l..mid]
# 2nd subarray is arr[mid+1..r]


def merge(arr, l, mid, r):
    n1 = mid - l + 1
    n2 = r - mid

    # create temp arrays
    Left = [0] * (n1)
    Right = [0] * (n2)

    # Copy arr to temp arrays L[] and R[]
    for i in range(0, n1):
        Left[i] = arr[l + i]

    for j in range(0, n2):
        Right[j] = arr[mid + 1 + j]

    # Merge the temp arrays back into original array arr[l..r]
    i = 0     # 1st index of first subarray
    j = 0     # 1st index of second subarray
    k = l     # 1st index of merged subarray

    while i < n1 and j < n2:
        if Left[i] <= Right[j]:
            arr[k] = Left[i]
            i += 1
        else:
            arr[k] = Right[j]
            j += 1
        k += 1

    # Copy the remaining elements of L[], if there are any
    while i < n1:
        arr[k] = Left[i]
        i += 1
        k += 1

    # Copy the remaining elements of R[], if there are any
    while j < n2:
        arr[k] = Right[j]
        j += 1
        k += 1


def mergeSort(arr, l, r):
    if l < r:

        # Same as (l+r)//2, but we avoid overflow for large l and h
        mid = (l+(r-1))//2

        # Sort first and second half
        mergeSort(arr, l, mid)
        mergeSort(arr, mid+1, r)
        merge(arr, l, mid, r)


# Driver code
arr = [12, 11, 13, 5, 6, 7]
n = len(arr)
print("Input array:")
for i in range(n):
    print("%d" % arr[i]),

mergeSort(arr, 0, n-1)
print("\n\nSorted array:")
for i in range(n):
    print("%d" % arr[i]),

# Sample input and output 1
# Input array:
# 12 11 13 5 6 7

# Sorted array:
# 5 6 7 11 12 13

# Sample input and output 2
# Input array:
# 15 17 32 3 8 9

# Sorted array:
# 3 8 9 15 17 32

# Time complexity : O(nlog(n))
# Space complexity : O(n)
